crates/shared/src/price_estimation/sanitized.rs

```diff
@@ -0,0 +1,313 @@
+        }
+        token_quality_errors
+    }
+
+    async fn bulk_estimate_prices(
+        &self,
+        queries: &[Query],
+    ) -> (Vec<EstimationProgress>, Vec<EstimationResult>) {
+        let token_quality_errors = self.get_token_quality_errors(queries).await;
+
+        let mut estimations_to_forward = Vec::new();
+
+        let estimation_progress = queries
+            .iter()
+            .map(|query| {
+                if let Some(err) = token_quality_errors.get(&query.buy_token) {
+                    return EstimationProgress::TrivialSolution(Err(err.clone()));
+                }
+                if let Some(err) = token_quality_errors.get(&query.sell_token) {
+                    return EstimationProgress::TrivialSolution(Err(err.clone()));
+                }
+
+                if query.buy_token == query.sell_token {
+                    return EstimationProgress::TrivialSolution(Ok(Estimate {
+                        out_amount: query.in_amount,
+                        gas: 0.into(),
+                    }));
+                }
+
+                if query.buy_token == BUY_ETH_ADDRESS {
+                    estimations_to_forward.push(Query {
+                        buy_token: self.native_token,
+                        ..*query
+                    });
+                    return EstimationProgress::AwaitingEthEstimation;
+                }
+
+                estimations_to_forward.push(*query);
+                EstimationProgress::AwaitingErc20Estimation
+            })
+            .collect::<Vec<_>>();
+
+        let remaining_estimations = self.inner.estimates(&estimations_to_forward[..]).await;
+        (estimation_progress, remaining_estimations)
+    }
+
+    fn merge_trivial_and_forwarded_estimations(
+        &self,
+        partially_estimated: Vec<EstimationProgress>,
+        forwarded_estimations: Vec<EstimationResult>,
+    ) -> Vec<EstimationResult> {
+        use EstimationProgress::*;
+
+        let mut forwarded_estimations = forwarded_estimations.into_iter();
+
+        partially_estimated
+            .into_iter()
+            .map(|progress| match progress {
+                TrivialSolution(res) => res,
+                AwaitingErc20Estimation => forwarded_estimations
+                    .next()
+                    .expect("there is a result for every forwarded estimation"),
+                AwaitingEthEstimation => {
+                    let mut res = forwarded_estimations
+                        .next()
+                        .expect("there is a result for every forwarded estimation")?;
+                    res.gas =
+                        res.gas
+                            .checked_add(GAS_PER_WETH_UNWRAP.into())
+                            .ok_or(anyhow::anyhow!(
+                                "cost of unwrapping ETH would overflow gas price"
+                            ))?;
+                    Ok(res)
+                }
+            })
+            .collect()
+    }
```

Comment:
Should we `debug_assert!(forwarded_estimatesion.is_empty())` here for good measure? Might help catch some logic errors in unit tests.

Reply:
Good idea!
`is_empty()` lead to following error:
```
[rustc E0658] [E] use of unstable library feature 'exact_size_is_empty'
see issue #35428 <https://github.com/rust-lang/rust/issues/35428> for more information
```
So I went with `debug_assert!(forwarded_estimations.next().is_none())`.

Reply:
My suggestions are usually half-pseudo-code 😛

> is_empty() lead to following error

Makes sense, its a `vec::IntoIter` and not a `Vec`. Cool that they have a proposal for `is_empty` for `ExactSizeIterator`s though!